### PR TITLE
xslt: reduce repeated XPath work during Office startup

### PR DESCRIPTION
**Pull request by ttv20**
## Summary

- precompile dynamic `name` and `namespace` attribute value templates for `xsl:element` and `xsl:attribute`
- reuse libxslt's stylesheet-owned compiled XPath representation instead of recompiling expressions for every source node
- reuse bounded XPath evaluation contexts, value stacks, result objects, and short string buffers
- use validated compiled-instruction metadata before repeating XSLT namespace and local-name comparisons
- skip local-variable lookup when the local scope is empty
- avoid hashing variable names that are already owned by the transform dictionary
- keep dynamic `xsl:element` namespace values alive through namespace creation instead of interning a transient copy
- grow the retained XPath string buffer geometrically while evaluating multi-argument `concat()`
- add public-API regression coverage for repeated dynamic element and attribute transformations

## Performance evidence

Two readiness-bounded LBR profiles per side measured the AVT precompile:

| Two-run mean | Before | After | Change |
|---|---:|---:|---:|
| Click-to-Run weighted cycles | 7.124B | 5.578B | -21.70% |
| `xsltApplySequenceConstructor` cycles | 3.570B | 2.075B | -41.87% |
| Heap/allocation leaf cycles | 2.679B | 1.808B | -32.51% |
| Profiled readiness window | 17.870 s | 16.651 s | -6.82% |

`xmlXPathCtxtCompile` and `xsltAttrTemplateValueProcessNode` disappeared from both candidate profiles. Required XPath evaluation remains under `xsltEvalAVT`.

For the later runtime changes:

- a repeated-XPath microbenchmark improved from a 173.405 ms median to 169.812 ms, about 2.1%
- allocation calls fell from 1,000,099 to 400,099; the current short-string pool trades part of that saving for more `realloc` calls
- the compiled-instruction dispatch removed about 42.1M cycles of repeated sequence/choose classification, 0.82% of the reference Click-to-Run cycles
- the empty-local-scope guard removed a sampled 10.224M-cycle dictionary lookup path; `xmlDictHashName` fell from 96.682M to 59.839M inclusive cycles in the next profile
- the dictionary-ownership guard removed the remaining variable-lookup hash path: 42.027M sampled cycles before, zero after; the only remaining `xmlDictHashName` samples came from dynamic `xsl:element` namespaces
- across two follow-up profiles, Click-to-Run weighted cycles averaged 4.400B versus 4.521B in the immediately preceding profile (-2.67%); readiness wall time remained noisy and regressed overall, so the eliminated call path is the stronger evidence
- retaining the evaluated namespace value reduced `xmlDictHashName` from a 49.852M-cycle two-profile mean to 19.372M cycles; the remaining path interns dynamic element names, whose lifetime must match the result document
- total Click-to-Run cycles were effectively unchanged for that last step (4.400B before and 4.400B after), so it should be treated as a small call-path cleanup of about 0.7%, not a startup-time claim
- geometric concatenation removed the 75.775M-cycle `xmlStrncat` path and reduced `xmlXPathConcatFunction` from 119.123M cycles to a 63.483M two-profile mean (-46.7%); total Click-to-Run cycles did not improve reliably in those samples

The removed call paths are the primary evidence. Whole-startup time and total cycles vary between runs, so these percentages should not be added together.

### Combined end-to-end result with PR 71

Three clean prefix-matched pairs measured this PR together with the companion MSXML parser and buffering changes in [PR 71](https://github.com/ttv20/wine4office/pull/71):

| Metric | Baseline | Combined candidate | Change |
|---|---:|---:|---:|
| Useful-ready | 17.678 s | 12.431 s | -5.247 s (-29.7%) |
| Office launch to window | 15.845 s | 10.665 s | -5.180 s (-32.7%) |
| Peak RSS | 886.3 MiB | 853.0 MiB | -33.3 MiB (-3.8%) |
| Peak PSS | 650.9 MiB | 610.7 MiB | -40.2 MiB (-6.2%) |
| Peak private | 566.3 MiB | 526.0 MiB | -40.3 MiB (-7.1%) |
| OfficeClickToRun CPU to window | 6.923 s | 2.273 s | -4.650 s (-67.2%) |

The runner synchronizes the restored prefix's `.update-timestamp` with the selected `wine.inf`, rejects any run that invokes `InstallHinfSection`, and compares absolute monotonic timestamps from Office launch. This corrected two earlier measurement errors: a baseline-only prefix upgrade and a probe-relative clock that omitted about 2.1 seconds. The result measures both PRs together and is not an attribution to this PR alone. It does not include the later experimental wineserver PE-backing cache.

## Validation

- the exact PR branch built `msxml3.dll` and `msxml3_test.exe` independently for x86-64 and i386
- the `domdoc` suite passed on both architectures and exercises the XSLT path through public MSXML interfaces
- the dynamic XSLT test covers `name` and `namespace` AVTs on both `xsl:element` and `xsl:attribute`, including repeated use of one compiled instruction
- the test parses the result and checks local names, namespace URIs, and values instead of relying on serializer formatting
- the equivalent transformation passed with native MSXML 3 on the dedicated Windows endpoint
- the combined research runner also passed `saxreader`, `schema`, `xmldoc`, and `xmlparser` on x86-64 and i386 and launched Excel to useful UI
- after the dictionary and transient-namespace changes, all five MSXML suites (`domdoc`, `saxreader`, `schema`, `xmldoc`, and `xmlparser`) passed serially on both x86-64 and i386, and Excel again reached useful UI
- after the concat change, the five suites passed again on both architectures; the standalone 100,000-iteration XPath test passed concat, cache-toggle, and object-copy stress checks in five runs

### Review follow-up, 2026-09-06

- Stop dependent DOM calls after a failed result load, element lookup, or attribute lookup, including `S_FALSE`; release every acquired interface and string through cleanup.
- Initialize each repeated transform result and compare it only after success.
- Updated the branch with `origin/main` at `26cb4e6206ea` and added a changelog entry.
- Commit `2ecbc612fe5`: focused libxml2/libxslt, MSXML DLL and test builds passed for x86-64 and i386 with no compiler warnings. The `domdoc` suite passed serially on both architectures: 43,591 checks, 253 expected todo results, zero failures and zero skips per run.

## Risk notes

- the XPath cache is owned by the existing transform context; parser-context reuse is one-deep and retains only stacks of at most 40 entries
- the separate short-string pool is capped at 16 buffers of at most 1,024 bytes
- cached string capacity is kept in the secondary XPath index field, separate from the primary index used by XSLT sorting; generic object copies and sort-key replacement clear that cache-only metadata
- compiled dispatch accepts metadata only when it points back to the current instruction, has the expected type, and is not an extension marker; all other cases retain the existing checks
- DTD-provided attributes and instructions without compiled AVT metadata retain the existing runtime path
- dictionary ownership is checked before skipping interning; names from external dictionaries still take the original lookup and second-scan path
- the transient namespace string is freed only after namespace search/creation has duplicated or reused it; all error paths retain ownership cleanup
- concat growth checks integer overflow, reports allocation failure through the XPath context, and retains only buffers within the existing 1,024-byte cache limit

## Summary by Sourcery

Reduce repeated XSLT and XPath processing overhead during Office startup while adding regression coverage for dynamic transformations and failure handling.

Bug Fixes:
- Prevent failed XML transformation checks from crashing the test suite.

Enhancements:
- Speed up repeated XSLT and XPath evaluation during Office startup by reducing recompilation, allocations, variable-lookup overhead, and repeated instruction classification.
- Improve XPath object, evaluation-context, string-buffer, and dynamic namespace reuse while preserving bounded caching and existing fallback behavior.

Tests:
- Add MSXML public-API regression coverage for dynamic XSLT element and attribute names, namespaces, transformation failures, and repeated XPath evaluation.

## Summary by CodeRabbit

- **Performance Improvements**
  - Improved performance for repeated XSLT and XPath evaluations during Office startup.
  - Reduced overhead when reusing XPath expressions, values, and string results.

- **Bug Fixes**
  - Improved support for dynamically generated XSLT element and attribute names, including namespaces.
  - Improved reliability for repeated transformations involving variables, nested functions, and sorting.
  - Added handling for compiled XSLT instructions and dynamic attribute-value templates.
  - Failed XML transformation checks are now reported without crashing the test suite.

### CODE REVIEW

**Comment by ttv20** [comment]
ChatGPT 5.6-sol Web:

I reviewed the XPath/XSLT caching changes and did not find a concrete correctness bug in the paths I checked. In particular, the `xsltGetNsPropNode` behavior initially looked suspicious for unqualified attributes on XSLT elements, but that matches libxslt's existing `xsltGetNsProp()` semantics, so I don't think that part is wrong.

The main concern here is regression surface and API layering. This PR changes several hot libxml2/libxslt internals at once: precompiled AVTs, reusable XPath parser/evaluation state, object/string caches, variable lookup shortcuts, and compiled-instruction metadata. The logic looks internally consistent, but this is broad enough that Office/MSXML tests alone may not exercise the edge cases.

One maintainability point: `xmlXPathReleaseObject()` and `xmlXPathCacheObjectCopy()` are being promoted from internal/static helpers into `xpath.h` as `XMLPUBFUN` APIs to support libxslt. If possible, I would keep these in an internal/private header instead of expanding libxml2's public API for what is effectively libxslt-private coordination.

My recommendation would be to keep this draft until it has broader generic libxml2/libxslt XPath/XSLT coverage (especially nested evaluations, variables, sorting/node-set replacement, AVTs, and error paths). I don't see a blocker in the code as written, but this one deserves more regression testing than the smaller PRs.

**Comment by ttv20** [comment]
@sourcery-ai review

**Comment by ttv20** [comment]
@coderabbitai review

**Comment by ttv20** [comment]
Review follow-up for 5d85a8d20eb:

- FIXED (ff0031d0850): moved xmlXPathReleaseObject() and xmlXPathCacheObjectCopy() out of the public libxml/xpath.h API and into the private libxml2 XPath header used by the two libxslt consumers.
- FIXED (5d85a8d20eb): expanded public MSXML coverage for cached string, number, boolean, and node-set variables; local/global shadowing; nested XPath functions; sorting and sort-key replacement; 16 repeated transforms; and AVT error cleanup.
- Native evidence: MSXML 3.0 and 6.0 both return E_FAIL for a dynamic QName with an undeclared prefix. Wine currently returns S_OK with an empty result, so the test records the native contract as todo_wine while exercising the cleanup path.
- Validation: focused libxml2, libxslt, msxml3.dll, and msxml3_test.exe builds passed for i386 and x86_64. The domdoc suite passed serially on both architectures. Parallel execution was discarded because existing leading_spaces.xml tests shared a temporary directory and collided.

The PR remains draft because CodeRabbit and Sourcery both rate-limited the requested review of the latest SHA; no automated findings were produced.

File: dlls/msxml3/tests/domdoc.c [hunk 1]
```diff
@@ -14135,10 +14135,149 @@ static const char omitxmldecl_doc[] =
...
+    BSTR str;
+
+    doc = create_document(&IID_IXMLDOMDocument);
+    hr = IXMLDOMDocument_loadXML(doc, result, &b);
+    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
+    ok(b == VARIANT_TRUE, "Failed to load transform result %s.\n", wine_dbgstr_w(result));
+
+    hr = IXMLDOMDocument_get_documentElement(doc, &root);
+    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
+    hr = IXMLDOMElement_get_childNodes(root, &children);
```
[the remark this answers is not in the document]
Comment (ttv20) [reply dlls/msxml3/tests/domdoc.c:14399]: FIXED in 2ecbc612fe5. Added S_OK and pointer guards before dependent DOM operations, centralized interface/string cleanup, guarded the new XML loads, and initialized each transform result before comparing it only on success. Focused builds and domdoc passed on x86-64 and i386: 43,591 checks per run, zero failures.

**Comment by ttv20** [comment]
FIXED in 2ecbc612fe5ba9c72d350dd2d92ba211cc99ce4e.

The dynamic-result helper now stops on failed loads and unsuccessful DOM lookups, including S_FALSE, before using the returned interfaces. Its cleanup releases all acquired interfaces and strings. The repeated-transform loop initializes ret on every iteration and compares it only after S_OK. The new stylesheet/source loads are guarded as well.

Focused builds passed without compiler warnings on x86-64 and i386. The domdoc suite passed serially on each architecture with 43,591 checks, 253 expected todo results, zero failures and zero skips. The branch includes current origin/main at 26cb4e6206ea.

@coderabbitai review
@sourcery-ai review

**Comment by ttv20** [comment]
@greptileai Please review the current head 2ecbc612fe5ba9c72d350dd2d92ba211cc99ce4e, including the XSLT test failure-path fix.

**Comment by ttv20** [comment]
Review follow-up complete at 2ecbc612fe5ba9c72d350dd2d92ba211cc99ce4e.

- FIXED: the XSLT test failure-path finding; the thread is resolved.
- Validation: focused x86-64/i386 builds passed without compiler warnings. Both serial domdoc runs passed with 43,591 checks, 253 expected todo results, zero failures, and zero skips each.
- CodeRabbit, Sourcery, and Greptile all completed successfully on this SHA. CodeRabbit confirmed the fix, Sourcery approved, and Greptile reviewed 9 files with no comments.
- The review watcher settled successfully. No outstanding findings remain; the PR is ready for the maintainer's merge.

**Comment by ttv20** [comment]
Updated with origin/main fb7dfec7138 before the requested merge. Resolved the CHANGELOG-only conflict by preserving both entries. The final head builds successfully for both architectures; domdoc, saxreader, schema, xmldoc, and xmlparser all pass serially on x86-64 and i386 with zero failures and zero skips. @greptileai Please review the updated head.
